### DISCUSSION TITLE
Access current trip/coordinates from the TripLayers class

### DISCUSSION THREAD

**Discussion by charlieforward9** [visgl/deck.gl#7549] [Q&A] [answered]
I would like to be able to pan to the coordinate on the map where the current trip is being drawn, and make it smooth. Currently, I have hardcoded the values in using currentTime and my own knowledge of where I was at the time, but I would like to have a more robust and less cluttered way of doing this.

I have been using the tutorials and reading all the docs, and have been trying to solve this for 2 days. So far nothing stood out to me, but I may be missing a simple fix. I imagine its not as simple as it sounds, as there can be multiple trips being drawn at the same time, but my use case does not do this, so I figure there is either a way to do it, or there should be a way! [FEATURE?]

 Please help!

My current implementation is as follows:

```typescript
// import { GoogleMapsOverlay } from "@deck.gl/google-maps";
// import { TripsLayer } from "deck.gl";

const GoogleMapsOverlay = deck.GoogleMapsOverlay;
const TripsLayer = deck.TripsLayer;

interface Data {
  path: [number, number][];
  timestamps: number[];
}

const DATA_URL =
  "https://raw.githubusercontent.com/charlieforward9/animated_heatmap/master/data/2022.json";

const LOOP_LENGTH = 31557600;

function initMap(): void {
  const map = new google.maps.Map(
    document.getElementById("map") as HTMLElement,
    {
      center: { lat: 29.64462421696083, lng: -82.33479384825146},
      mapId: '1ae1962daafbdd69',
      tilt: 45,
      zoom: 12,
      disableDefaultUI: true,
    } as google.maps.MapOptions
  );
  const view = new deck.MapView({id:"view", x:29.64462421696083, y: -82.33479384825146, width: 300, height: 200});

  let currentTime = -1000000;
  let playSpeed = 30000;
  const props = {
    id: "trips",
    data: DATA_URL,
    getPath: (d: Data) =>d.path,
    getTimestamps: (d: Data) => d.timestamps,
    getColor: [255, 87, 51],
    opacity: 1,
    widthMinPixels: 4,
    trailLength: 31557600,
    currentTime,
    shadowEnabled: false,
    jointRounded: true,
    capRounded: true
  };

  //Hardcoded panning (WHAT I AM TRYING TO AVOID)
  function autoAnimate(): void {
    window.setInterval(viewportAnimation, 80);
  }
  function viewportAnimation(): void {
    const heading = map.getHeading() || 0;
    const zoom = map.getZoom() || 10;
    const tilt = map.getTilt() || 45;
  
    console.log(currentTime);

    if (currentTime < 3600000) {
      //Gainesville
      map.setHeading(heading + 0.1);
      map.setZoom(zoom - 0.01);

    } else if (currentTime < 4000000) {
      //Orlando SoFlo
      map.setHeading(0);
      map.setZoom(zoom + 0.1);
      map.setTilt(tilt - 0.1);
      map.panTo({ lat: 28.6024, lng: -81.2001});

    } else if (currentTime < 7300000) {
      //Gainesville 
      map.setHeading(heading - 0.1);
      map.setZoom(zoom - 0.01);
      map.panTo({ lat: 29.64462421696083, lng: -82.33479384825146});
      
    } 
  ...
  }

  const overlay = new GoogleMapsOverlay({});

  const animate = () => {
    currentTime = (currentTime + playSpeed) % LOOP_LENGTH;

    const tripsLayer = new TripsLayer({
      ...props,
      currentTime,
    });

    overlay.setProps({
      layers: [tripsLayer],
    });

    window.requestAnimationFrame(animate);
  };

  window.requestAnimationFrame(animate);
  autoAnimate()

  overlay.setMap(map);
}

declare global {
  interface Window {
    initMap: () => void;
  }
}
window.initMap = initMap;
export {};
```

**Comment by Pessimistress** [comment, marked as answer]
The TripLayer does not interpolate the input data on the CPU. You will need to compute it yourself.

**Reply by charlieforward9** [reply]
Thank you for providing some feedback. This is unfortunate to hear. Is there any discussion about doing so? Either way, could you point me in the direction of some resources that could give me more context on this rationale? I am hoping to learn more about this, thank you.

**Reply by Pessimistress** [reply]
Computation on the CPU is expensive and slow. The point of the library is to leverage the GPU for better performance. I'm not sure what your expectation is.

**Reply by charlieforward9** [reply]
Got it. Thanks. I am still very new to these subjects, so just asking questions I know how to ask now in hopes of learning more little by little. Appreciate the responses.

**Reply by Pessimistress** [reply]
You seem to imply that the layer holds the current location of each trip's head in memory somewhere, and just has to expose it. That is not true. The interpolation is done entirely on the GPU, and the CPU has no visibility into where that head pixel is.

You can look at [this example](https://deck.gl/examples/google-maps/) which implements exactly what you described.

https://github.com/visgl/deck.gl/blob/master/examples/website/google-3d/trip-builder.js#L71

**Reply by charlieforward9** [reply]
That is what I was hoping, but I was clearly incorrect, your rationale makes complete sense. Thank you for finding an example solution to my description! You have been very accommodating in this discussion.

I look forward to implementing this in my project and continuing to use the DeckGL library. I appreciate the open-source contribution you make and see you have been doing it for quite some time. This is very inspiring. I will try to pay it forward as an aspiring designer-developer.
